src/power_grid_model/core/power_grid_core.py

```diff
@@ -0,0 +1,220 @@
+    set_threading: Callable[[OptionsPtr, int], None]
+    # model
+    create_model: Callable[[float, int, CharDoublePtr, IdxPtr, VoidDoublePtr], ModelPtr]  # type: ignore
+    update_model: Callable[[ModelPtr, int, CharDoublePtr, IdxPtr, VoidDoublePtr], None]  # type: ignore
+    copy_model: Callable[[ModelPtr], ModelPtr]
+    get_indexer: Callable[[ModelPtr, str, int, IDPtr, IdxPtr], None]  # type: ignore
+    calculate: Callable[  # type: ignore
+        [
+            # model
+            ModelPtr,
+            OptionsPtr,
+            # output
+            int,
+            CharDoublePtr,  # type: ignore
+            VoidDoublePtr,  # type: ignore
+            # update
+            int,
+            int,
+            CharDoublePtr,  # type: ignore
+            IdxPtr,
+            IdxDoublePtr,  # type: ignore
+            VoidDoublePtr,  # type: ignore
+        ],
+        None,
+    ]
+    destroy_model: Callable[[ModelPtr], None]
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls, *args, **kwargs)
+        instance._cdll = _load_core()
+        instance._handle = instance._cdll.PGM_create_handle()
+        return instance
+
+    def __init__(self):
+        for name, function in PowerGridCore.__annotations__.items():  # pylint: disable=E1101
+            if name.startswith("_"):
+                continue
+            # get and convert types
+            py_argtypes = function.__args__[:-1]
+            py_restype = function.__args__[-1]
+            c_argtypes = [_ARGS_TYPE_MAPPING.get(x, x) for x in py_argtypes]
+            c_restype = _ARGS_TYPE_MAPPING.get(py_restype, py_restype)
+            if c_restype == Idx_c and name in _FUNC_SIZE_T_RES:
+                c_restype = c_size_t
+            # bug in Python 3.10 https://bugs.python.org/issue43208
+            if id(c_restype) == id(type(None)):
+                c_restype = None
+            # set argument in dll
+            # mostly with handle pointer, except destroy function
+            is_destroy_func = "destroy" in name
+            if is_destroy_func:
+                getattr(self._cdll, f"PGM_{name}").argtypes = c_argtypes
+            else:
+                getattr(self._cdll, f"PGM_{name}").argtypes = [HandlePtr] + c_argtypes
+            getattr(self._cdll, f"PGM_{name}").restype = c_restype
+            # set wrapper functor to instance
+            setattr(
+                self,
+                name,
+                WrapperFunc(
+                    cdll=self._cdll, handle=self._handle, name=name, c_argtypes=c_argtypes, c_restype=c_restype
+                ),
+            )
+
+    def __del__(self):
+        self._cdll.PGM_destroy_handle(self._handle)
+
+    # not copyable
+    def __copy__(self):
+        raise NotImplementedError("Class not copyable")
+
+    def __deepcopy__(self, memodict):
+        raise NotImplementedError("class not copyable")
+
+
+# make one instance
+power_grid_core = PowerGridCore()
```

Comment:
`PowerGridCore` should be private, can we call it `_PowerGridCore`? Is that a valid python name? or we could rename the file to _power_grid_code.py and create the instance in the `__init__.py`. But, to be honest, I'm not a big fan of these one-instance setups, is it get's instantiated on import. I'd prefer a singleton construction, e.g. something like this:

```python
class PowerGridCore:
    _instance: Optional["PowerGridCore"] = None

def __new__():
    if PowerGridCore._instance is None:
        PowerGridCore._instance = super().__new__()
    return PowerGridCore._instance
```